File: sdk/src/quote.ts
/**
 * A client-side swap quote.
 *
 * This is a line-by-line mirror of `instructions::swap` and `math::swap`, in
 * `bigint`. It exists because `min_amount_out` is the only thing standing
 * between a trader and a fill they did not want, and a slippage bound computed
 * from a guess is worse than none — it either rejects good fills or permits
 * bad ones.
 *
 * Mirroring means mirroring the rounding. Outputs floor and required inputs
 * ceil, always in the pool's favour; the volatility accumulator is advanced
 * per bin crossed, exactly as the walk does, because under a taper the fee is
 * a function of the bin being crossed rather than a pool constant.
 *
 * The quote is exact for the state it was given. It is still only a quote:
 * another trade landing first moves the active bin, which is what slippage
 * tolerance is for.
 */
import { FEE_PRECISION, MAX_FEE_RATE, ONE_Q64 } from "./constants";
import { binArrayIndex } from "./pda";
import { feeRateForStep } from "./ladder";
import type { BinView, ConfigView, PoolView } from "./types";

/** `MAX_BINS_PER_SWAP` in the program. */
export const MAX_BINS_PER_SWAP = 200;

const U64_MAX = (1n << 64n) - 1n;
const PRECISION = BigInt(FEE_PRECISION);

const clampU64 = (v: bigint) => (v > U64_MAX ? U64_MAX : v);
const ceilDiv = (a: bigint, b: bigint) => (a + b - 1n) / b;

/** `ceil(amount * rate / 1e9)` — a fee carved out of an amount. */
const feeFromAmount = (amount: bigint, rate: bigint) =>
  rate === 0n ? 0n : clampU64(ceilDiv(amount * rate, PRECISION));

/** `ceil(amount * rate / (1e9 - rate))` — a fee added on top of a net amount. */
const feeOnAmount = (amount: bigint, rate: bigint) =>
  rate === 0n ? 0n : clampU64(ceilDiv(amount * rate, PRECISION - rate));

const outputFor = (netIn: bigint, priceQ64: bigint, swapForY: boolean) =>
  swapForY ? (netIn * priceQ64) / ONE_Q64 : (netIn * ONE_Q64) / priceQ64;

export type BinSwap = {
  amountIn: bigint;
  amountOut: bigint;
  fee: bigint;
  binIn: bigint;
  binOut: bigint;
};

const NOOP: BinSwap = { amountIn: 0n, amountOut: 0n, fee: 0n, binIn: 0n, binOut: 0n };
const isNoop = (r: BinSwap) => r.amountIn === 0n && r.amountOut === 0n;

/** Mirrors `math::swap::swap_in_bin`. */
export function swapInBin(
  binAmountX: bigint,
  binAmountY: bigint,
  priceQ64: bigint,
  amountInLeft: bigint,
  swapForY: boolean,
  feeRate: bigint,
  feeOnOutput: boolean
): BinSwap {
  const maxBinOut = swapForY ? binAmountY : binAmountX;
  if (maxBinOut === 0n || amountInLeft === 0n || priceQ64 === 0n) return NOOP;

  // Input, net of any input-side fee, that would drain the bin. Saturating is
  // correct: a requirement past u64 cannot be met by `amountInLeft` anyway.
  const maxNetIn = clampU64(
    swapForY ? ceilDiv(maxBinOut * ONE_Q64, priceQ64) : ceilDiv(maxBinOut * priceQ64, ONE_Q64)
  );

  if (feeOnOutput) {
    // Quote-only collection on an X -> Y swap: the whole input enters the bin
    // and the fee is taken out of the Y leaving it.
    let amountIn: bigint;
    let binOut: bigint;
    if (amountInLeft >= maxNetIn) {
      amountIn = maxNetIn;
      binOut = maxBinOut;
    } else {
      amountIn = amountInLeft;
      const out = outputFor(amountInLeft, priceQ64, swapForY);
      binOut = out < maxBinOut ? out : maxBinOut;
    }
    if (binOut === 0n) return NOOP;
    const fee = feeFromAmount(binOut, feeRate);
    return { amountIn, amountOut: binOut - fee, fee, binIn: amountIn, binOut };
  }

  // Fee on the input side.
  const maxFee = feeOnAmount(maxNetIn, feeRate);
  const maxGrossIn = maxNetIn + maxFee;

  if (maxGrossIn <= U64_MAX && amountInLeft >= maxGrossIn) {
    return {
      amountIn: maxGrossIn,
      amountOut: maxBinOut,
      fee: maxFee,
      binIn: maxNetIn,
      binOut: maxBinOut
    };
  }

  const fee = feeFromAmount(amountInLeft, feeRate);
  const net = amountInLeft - fee;
  const raw = outputFor(net, priceQ64, swapForY);
  const out = raw < maxBinOut ? raw : maxBinOut;
  if (out === 0n) return NOOP;
  return { amountIn: amountInLeft, amountOut: out, fee, binIn: net, binOut: out };
}

export type QuoteInput = {
  pool: PoolView;
  config: ConfigView;
  /** Bins by id. Only bins the program has derived can trade. */
  bins: Map<number, BinView>;
  /** Whether the bin array covering an index was supplied to the swap. */
  hasArray: (index: number) => boolean;
  /** Input in *arrival* units — after the input mint's transfer fee. */
  amountIn: bigint;
  swapForY: boolean;
  /** Unix seconds the transaction will land at. Drives the volatility decay. */
  now: number;
};

/** One bin's contribution to a walk, in the order the walk visited them. */
export type BinFill = BinSwap & {
  binId: number;
  priceQ64: bigint;
  /** The rate this bin charged, against `FEE_PRECISION`. */
  feeRate: bigint;
};

/**
 * The pool fields the walk advanced.
 *
 * A quote is a pure function of the state it is handed, so simulating a
 * *sequence* of swaps — a round trip, a route, a backtest — means feeding this
 * back in along with the bin inventories `fills` moved. It is exactly what
 * `instructions::swap` leaves on the pool account.
 */
export type WalkState = {
  activeId: number;
  indexReference: number;
  volatilityAccumulator: number;
  volatilityReference: number;
  lastUpdateTimestamp: bigint;
};

export type Quote = {
  /** Consumed from the budget. Less than `amountIn` on a partial fill. */
  amountIn: bigint;
  /** Delivered from the reserves, before the output mint's transfer fee. */
  amountOut: bigint;
  /** Total swap fee, in whichever token the mode collects it. */
  fee: bigint;
  binsCrossed: number;
  startId: number;
  endId: number;
  /** True when the pool could not absorb the whole input. */
  partial: boolean;
  /** Average execution price, Y-lamports per X-lamport. */
  executionPrice: number;
  /** Fee as a fraction of the input actually consumed. */
  effectiveFeeRate: number;
  /** Per-bin detail, in visit order. */
  fills: BinFill[];
  /** Pool state after the walk. */
  state: WalkState;
};

/**
 * Walks the ladder the way the program does and reports what would come back.
 *
 * `hasArray` matters as much as the bins do: the program breaks its walk the
 * moment it needs an array that was not passed as a remaining account, so a
 * quote that ignored array availability would promise fills the transaction
 * cannot deliver. Use `swapArrayIndexes` to build both from the same source.
 */
export function quoteSwap({
  pool,
  config,
  bins,
  hasArray,
  amountIn,
  swapForY,
  now
}: QuoteInput): Quote {
  const startId = pool.activeId;
  let activeId = startId;
  let amountLeft = amountIn;
  let totalIn = 0n;
  let totalOut = 0n;
  let totalFee = 0n;
  let crossed = 0;
  const fills: BinFill[] = [];

  const feeOnOutput = config.collectFeeMode === 1 && swapForY;
  const step = swapForY ? -1 : 1;
  const inBand = (id: number) => id >= config.minBinId && id <= config.maxBinId;

  // `update_references`: decay the reference if the pool has been idle past
  // the filter period, then re-anchor it.
  let indexReference = pool.indexReference;
  let volatilityReference = pool.volatilityReference;
  const elapsed = now - Number(pool.lastUpdateTimestamp);
  if (elapsed >= config.filterPeriod) {
    indexReference = pool.activeId;
    volatilityReference =
      elapsed < config.decayPeriod
        ? Math.floor((pool.volatilityAccumulator * config.reductionFactor) / 10_000)
        : 0;
  }
  let volatilityAccumulator = pool.volatilityAccumulator;

  for (let i = 0; i < MAX_BINS_PER_SWAP; i += 1) {
    if (amountLeft === 0n) break;
    if (!inBand(activeId)) break;
    if (!hasArray(binArrayIndex(activeId))) break;

    const bin = bins.get(activeId);
    const available = bin ? (swapForY ? bin.amountY : bin.amountX) : 0n;

    let filled = false;
    let exhausted = false;

    if (!bin || available === 0n) {
      exhausted = true;
    } else {
      // `update_volatility_accumulator`, per bin crossed.
      volatilityAccumulator = Math.min(
        volatilityReference + Math.abs(activeId - indexReference) * 10_000,
        config.maxVolatilityAccumulator
      );
      const feeRate = BigInt(
        Math.min(feeRateForStep(bin.stepBpX100, config, volatilityAccumulator), MAX_FEE_RATE)
      );

      const result = swapInBin(
        bin.amountX,
        bin.amountY,
        bin.priceQ64,
        amountLeft,
        swapForY,
        feeRate,
        feeOnOutput
      );

      if (isNoop(result)) {
        // The bin holds inventory but the remaining input buys none of it.
        break;
      }

      amountLeft -= result.amountIn;
      totalIn += result.amountIn;
      totalOut += result.amountOut;
      totalFee += result.fee;
      crossed += 1;
      filled = true;
      fills.push({ ...result, binId: activeId, priceQ64: bin.priceQ64, feeRate });
    }

    if (!filled && !exhausted) break;
    if (amountLeft === 0n) break;

    const next = activeId + step;
    if (!inBand(next)) break;
    activeId = next;
  }

  const executionPrice =
    totalIn > 0n && totalOut > 0n
      ? swapForY
        ? Number(totalOut) / Number(totalIn)
        : Number(totalIn) / Number(totalOut)
      : 0;

  return {
    amountIn: totalIn,
    amountOut: totalOut,
    fee: totalFee,
    binsCrossed: crossed,
    startId,
    endId: activeId,
    partial: totalIn < amountIn,
    executionPrice,
    effectiveFeeRate: totalIn > 0n ? Number(totalFee) / Number(totalIn) : 0,
    fills,
    state: {
      activeId,
      indexReference,
      volatilityAccumulator,
      volatilityReference,
      // `update_references` stamps the pool whether or not anything filled.
      lastUpdateTimestamp: BigInt(Math.trunc(now))
    }
  };
}

/**
 * `min_amount_out` for a quote at a given slippage tolerance, in bps.
 *
 * The program checks this against what *reaches the trader's wallet*, so the
 * output mint's transfer fee has already to have been applied to `amountOut`
 * before it gets here.
 */
export const minOutFor = (amountOut: bigint, slippageBps: number) =>
  (amountOut * BigInt(10_000 - Math.max(0, Math.min(10_000, Math.round(slippageBps))))) / 10_000n;

File: jupiter/taper-amm-sdk/src/quote.rs
//! What a swap would do.
//!
//! This is `instructions::swap` with the token transfers taken out. Every
//! number it produces comes from `taper_core` — `swap_in_bin`, the ladder, the
//! fee schedule, the volatility accumulator — so the quote is not an
//! approximation of the program's arithmetic, it *is* the program's
//! arithmetic. The walk mutates its own copy of the bins as it goes, for the
//! same reason the program does: what the second bin can sell depends on what
//! the first one just bought.
//!
//! Two things a caller must not get wrong, both of which this module handles
//! rather than leaving to them:
//!
//! - **A quote is not a pure function of account state.** `update_references`
//!   decays the volatility reference off the wall clock, so a stale timestamp
//!   quotes a fee that is too high or too low. `now` is a parameter, not a
//!   read.
//! - **Transfer fees move both ends.** The ladder is budgeted in what
//!   *arrives*, so the input's fee comes off before the walk and the output's
//!   comes off after it. [`quote_exact_in`] reports the trade the way the
//!   trader's own wallet will see it.

use std::collections::BTreeMap;

use taper_core::constants::MAX_BINS_PER_SWAP;
use taper_core::math::ladder::bin_array_index;
use taper_core::math::swap::swap_in_bin;
use taper_core::state::{BinArray, Config, Pool};
use taper_core::CoreError;

use crate::token::TransferFeeQuote;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum QuoteError {
    /// The pool is disabled, or some arithmetic in the ladder gave up.
    Core(CoreError),
    /// A bin array the walk needed was not in the map it was handed.
    MissingBinArray(i32),
    /// The transfer-fee arithmetic overflowed a `u64`.
    TransferFeeOverflow,
    /// Nothing filled: no liquidity in reach, or an input too small to buy a
    /// single lamport.
    NoFill,
    /// The mints given are not this pool's pair.
    WrongMints,
}

impl From<CoreError> for QuoteError {
    fn from(error: CoreError) -> Self {
        QuoteError::Core(error)
    }
}

impl core::fmt::Display for QuoteError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            QuoteError::Core(e) => write!(f, "{e}"),
            QuoteError::MissingBinArray(i) => write!(f, "bin array {i} was not supplied"),
            QuoteError::TransferFeeOverflow => f.write_str("transfer fee arithmetic overflowed"),
            QuoteError::NoFill => f.write_str("no liquidity in reach filled this swap"),
            QuoteError::WrongMints => f.write_str("mints are not this pool's pair"),
        }
    }
}

impl std::error::Error for QuoteError {}

/// What the ladder itself did, in arrival units at both ends.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Walk {
    /// Consumed from the budget. Less than the budget on a partial fill.
    pub amount_in: u64,
    /// Taken out of the reserve, before the output mint's fee.
    pub amount_out: u64,
    /// The swap fee, in whichever token the collect mode denominates it.
    pub fee: u64,
    pub bins_crossed: u32,
    pub start_id: i32,
    pub end_id: i32,
    /// The ladder could not absorb the whole budget.
    pub partial: bool,
}

/// The same trade as the trader's wallet sees it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct SwapQuote {
    /// Leaves the wallet, transfer fee included.
    pub in_amount: u64,
    /// Reaches the wallet, transfer fee already taken off.
    pub out_amount: u64,
    /// The pool's swap fee. Not the mints' transfer fees.
    pub fee_amount: u64,
    /// True when the swap fee is denominated in Y rather than X.
    pub fee_in_y: bool,
    pub walk: Walk,
}

/// Walks the ladder exactly as `instructions::swap` does.
///
/// `arrays` is what a client fetched; the walk copies each one it enters and
/// leaves the caller's untouched, so the same map can serve many quotes. That
/// copy is why this takes `&BTreeMap` and not `&mut`: `Amm::quote` has only
/// `&self`.
///
/// `budget` is in *arrival* units — the input mint's transfer fee has already
/// come off. Use [`quote_exact_in`] to start from a wallet amount instead.
pub fn walk_ladder(
    pool: &Pool,
    config: &Config,
    arrays: &BTreeMap<i32, BinArray>,
    budget: u64,
    swap_for_y: bool,
    now: i64,
) -> Result<Walk, QuoteError> {
    pool.require_enabled()?;

    let mut pool = *pool;
    pool.update_references(config, now)?;

    let ladder = config.ladder();
    let fee_on_output = config.collect_fee_mode()?.fee_on_output(swap_for_y);
    let step: i32 = if swap_for_y { -1 } else { 1 };
    let start_id = pool.active_id;

    // Copied on first touch rather than up front: a walk that stays in one
    // array should not pay to clone the two beside it.
    let mut touched: BTreeMap<i32, BinArray> = BTreeMap::new();

    let (mut amount_left, mut total_in, mut total_out, mut total_fee) = (budget, 0u64, 0u64, 0u64);
    let mut crossed = 0u32;

    for _ in 0..MAX_BINS_PER_SWAP {
        if amount_left == 0 {
            break;
        }
        let bin_id = pool.active_id;
        if !config.contains_bin(bin_id) {
            break;
        }
        let index = bin_array_index(bin_id);
        // A gap ends the walk, exactly as `BinArrays::get` does on chain.
        if !touched.contains_key(&index) {
            let Some(array) = arrays.get(&index) else {
                break;
            };
            touched.insert(index, *array);
        }
        let array = touched.get_mut(&index).expect("just inserted");

        let (filled, exhausted) = {
            let bin = array.bin_mut(bin_id, &ladder)?;
            let available = if swap_for_y { bin.amount_y } else { bin.amount_x };
            if available == 0 {
                (false, true)
            } else {
                pool.update_volatility_accumulator(config, bin_id)?;
                let fee_rate = pool.fee_rate_for_step(config, bin.step_bp_x100)?;

                let result = swap_in_bin(
                    bin.amount_x,
                    bin.amount_y,
                    bin.price,
                    amount_left,
                    swap_for_y,
                    fee_rate,
                    fee_on_output,
                )?;

                if result.is_noop() {
                    // Inventory the remaining input cannot buy a lamport of.
                    (false, false)
                } else {
                    if swap_for_y {
                        bin.amount_x = bin
                            .amount_x
                            .checked_add(result.bin_in)
                            .ok_or(CoreError::MathOverflow)?;
                        bin.amount_y -= result.bin_out;
                    } else {
                        bin.amount_y = bin
                            .amount_y
                            .checked_add(result.bin_in)
                            .ok_or(CoreError::MathOverflow)?;
                        bin.amount_x -= result.bin_out;
                    }
                    amount_left -= result.amount_in;
                    total_in += result.amount_in;
                    total_out += result.amount_out;
                    total_fee += result.fee;
                    crossed += 1;
                    (true, false)
                }
            }
        };

        if !filled && !exhausted {
            break; // dust
        }
        if amount_left == 0 {
            break;
        }
        let next = bin_id.checked_add(step).ok_or(CoreError::BinIdOutOfRange)?;
        if !config.contains_bin(next) {
            break;
        }
        pool.active_id = next;
    }

    Ok(Walk {
        amount_in: total_in,
        amount_out: total_out,
        fee: total_fee,
        bins_crossed: crossed,
        start_id,
        end_id: pool.active_id,
        partial: total_in < budget,
    })
}

/// A quote denominated at the trader's wallet on both ends.
///
/// `in_amount` is what leaves it and `out_amount` what reaches it, which is
/// where Jupiter measures and where the program's own `min_amount_out` guard
/// applies. Between them sits the ladder, which never sees either number.
///
/// The `min(amount_in)` on the way out is not defensive: `to_send` rounds
/// towards the reserve and can name one lamport more than the trader offered,
/// and the program caps it the same way for the same reason.
#[allow(clippy::too_many_arguments)]
pub fn quote_exact_in(
    pool: &Pool,
    config: &Config,
    arrays: &BTreeMap<i32, BinArray>,
    amount_in: u64,
    swap_for_y: bool,
    in_fee: &TransferFeeQuote,
    out_fee: &TransferFeeQuote,
    now: i64,
) -> Result<SwapQuote, QuoteError> {
    let budget = in_fee
        .received(amount_in)
        .ok_or(QuoteError::TransferFeeOverflow)?;
    if budget == 0 {
        return Err(QuoteError::NoFill);
    }

    let walk = walk_ladder(pool, config, arrays, budget, swap_for_y, now)?;
    if walk.amount_out == 0 {
        return Err(QuoteError::NoFill);
    }

    let in_amount = in_fee
        .to_send(walk.amount_in)
        .ok_or(QuoteError::TransferFeeOverflow)?
        .min(amount_in);
    let out_amount = out_fee
        .received(walk.amount_out)
        .ok_or(QuoteError::TransferFeeOverflow)?;

    // The fee lands in Y unless it was taken from an X input — the same
    // condition `instructions::swap` calls `fee_is_y`.
    let fee_on_output = config.collect_fee_mode()?.fee_on_output(swap_for_y);
    Ok(SwapQuote {
        in_amount,
        out_amount,
        fee_amount: walk.fee,
        fee_in_y: fee_on_output || !swap_for_y,
        walk,
    })
}
